pkg/observable/channel/internals.go

```diff
@@ -0,0 +1,111 @@
+package channel
+
+import (
+	"sync"
+
+	"pocket/pkg/observable"
+)
+
+// observableInternals is an interface intended to be used between an observable and some
+// higher-level abstraction and/or observable implementation which would embed it.
+// Embedding this interface rather than a channelObservable directly allows for
+// more transparency and flexibility in higher-level code.
+// NOTE: this interface MUST be used with a common concrete Observer type.
+type observableInternals[V any] interface {
+	addObserver(toAdd observable.Observer[V])
+	onUnsubscribe(toRemove observable.Observer[V])
+	unsubscribeAll()
+}
+
+type channelObservableInternals[V any] struct {
+	// observersMu protects observers from concurrent access/updates
+	observersMu *sync.RWMutex
+	// observers is a list of channelObservers that will be notified when publishCh
+	// receives a new value.
+	observers []*channelObserver[V]
+}
+
+func newObservableInternals[V any]() channelObservableInternals[V] {
+	return channelObservableInternals[V]{
+		observersMu: &sync.RWMutex{},
+		observers:   []*channelObserver[V]{},
+	}
+}
+
+// addObserver implements the respective member of observableInternals. It is used
+// by the channelObservable implementation as well as embedders of observableInternals
+// (e.g. replayObservable).
+// It panics if toAdd is not a channelObserver.
+func (coi *channelObservableInternals[V]) addObserver(
+	toAdd observable.Observer[V],
+) {
+	// must (write) lock observersMu so that we can safely append to the observers list
+	coi.observersMu.Lock()
+	defer coi.observersMu.Unlock()
+
+	coi.observers = append(coi.observers, toAdd.(*channelObserver[V]))
+}
+
+// onUnsubscribe returns a function that removes a given observer from the
+// observable's list of observers.
+// It implements the respective member of observableInternals and is used by
+// the channelObservable implementation as well as embedders of observableInternals
+// (e.g. replayObservable).
+func (coi *channelObservableInternals[V]) onUnsubscribe(
+	toRemove observable.Observer[V],
+) {
+	// must (write) lock to iterate over and modify the observers list
+	coi.observersMu.Lock()
+	defer coi.observersMu.Unlock()
+
+	for i, observer := range coi.observers {
+		if observer == toRemove {
+			coi.observers = append((coi.observers)[:i], (coi.observers)[i+1:]...)
```

Comment:
Double checking that we don't need to `Clean/Close/Shutdown/Unsubscribe` the `toRemove` observer here. Correct?

Reply:
Correct, that's outside the scope of this object.